edb/server/protocol/auth_ext/base.py

```diff
@@ -37,14 +43,98 @@ def __init__(
+    async def exchange_code(self, code: str) -> data.OAuthAccessTokenResponse:
         raise NotImplementedError
 
-    async def fetch_user_info(self, token: str) -> data.UserInfo:
+    async def fetch_user_info(
+        self, token_response: data.OAuthAccessTokenResponse
+    ) -> data.UserInfo:
         raise NotImplementedError
 
     def _maybe_isoformat_to_timestamp(self, value: str | None) -> float | None:
         return datetime.fromisoformat(value).timestamp() if value else None
+
+
+class OpenIDProvider(BaseProvider):
+    def __init__(self, name: str, issuer_url: str, *args, **kwargs):
+        super().__init__(name, issuer_url, *args, **kwargs)
+
+    async def get_code_url(self, state: str, redirect_uri: str) -> str:
+        oidc_config = await self._get_oidc_config()
+        params = {
+            "client_id": self.client_id,
+            "scope": "openid profile email",
+            "state": state,
+            "redirect_uri": redirect_uri,
+            "nonce": str(uuid.uuid4()),
+            "response_type": "code",
+        }
+        encoded = urllib.parse.urlencode(params)
+        return f"{oidc_config.authorization_endpoint}?{encoded}"
+
+    async def exchange_code(
+        self, code: str
+    ) -> data.OpenIDConnectAccessTokenResponse:
+        oidc_config = await self._get_oidc_config()
+
+        token_endpoint = urllib.parse.urlparse(oidc_config.token_endpoint)
+        async with self.http_factory(
+            base_url=f"{token_endpoint.scheme}://{token_endpoint.netloc}"
+        ) as client:
+            resp = await client.post(
+                token_endpoint.path,
+                json={
+                    "grant_type": "authorization_code",
+                    "code": code,
+                    "client_id": self.client_id,
+                    "client_secret": self.client_secret,
+                },
+            )
+            json = resp.json()
+
+            return data.OpenIDConnectAccessTokenResponse(**json)
+
+    async def fetch_user_info(
+        self, token_response: data.OAuthAccessTokenResponse
+    ) -> data.UserInfo:
+        if not isinstance(
+            token_response, data.OpenIDConnectAccessTokenResponse
+        ):
+            raise TypeError(
+                "token_response must be of type "
+                "OpenIDConnectAccessTokenResponse"
+            )
+        id_token = token_response.id_token
+
+        # Retrieve JWK Set
+        oidc_config = await self._get_oidc_config()
+        jwks_uri = urllib.parse.urlparse(oidc_config.jwks_uri)
+        async with self.http_factory(
+            base_url=f"{jwks_uri.scheme}://{jwks_uri.netloc}"
+        ) as client:
+            r = await client.get(jwks_uri.path)
+
+        # Load the token as a JWT object and verify it directly
+        jwk_set = jwk.JWKSet.from_json(r.text)
+        id_token_verified = jwt.JWT(key=jwk_set, jwt=id_token)
+        payload = json.loads(id_token_verified.claims)
```

Comment:
We should wrap the possible parsing errors here instead of propagating and triggering ISEs. The same for the following key references.

Reply:
I think one try/except is enough. It'll be all the fault of the provider, right? For debugging, we still have the traceback.

Reply:
> It'll be all the fault of the provider, right?

Yeah, I believe so. If the first function fails, it's because their `jwks_uri` isn't return a valid `JWKSet`, if the `JWT` instantiation fails it'll be due to a malformed `id_token`, and if the JSON parsing fails, it'll be due to a malformed JWT claim, which _might_ throw in the function before it since it will attempt to check expiration, etc. so it probably does some JSON parsing of it's own?

Reply:
Yeah, `jwt._check_provided_claims()` decoded the claim string and abandoned the parsed value.

Reply:
@fantix ⚡ [`00ec539` (#6013)](https://github.com/edgedb/edgedb/pull/6013/commits/00ec5391fa86b68c669e76d49ee7a95abb849f1a)

Does this seem about right?

Reply:
Awesome!